components/relevancy/src/populate_interests.rs

```diff
@@ -2,69 +2,121 @@
+use base64::{engine::general_purpose::STANDARD, Engine};
+use remote_settings::{Client, RemoteSettingsConfig, RemoteSettingsServer};
 
 // Number of rows to write when inserting interest data before checking for interruption
-const WRITE_CHUNK_SIZE: usize = 100;
+//const WRITE_CHUNK_SIZE: usize = 100;
 
 pub fn ensure_interest_data_populated(db: &RelevancyDb) -> Result<()> {
     if !db.read(|dao| dao.need_to_load_url_interests())? {
         return Ok(());
     }
-    let interest_data = match fetch_interest_data() {
-        Ok(data) => data,
+
+    match fetch_interest_data() {
+        Ok(data) => {
+            db.read_write(move |dao| insert_interest_data(data, dao))?;
+        }
         Err(e) => {
             log::warn!("error fetching interest data: {e}");
             return Err(Error::FetchInterestDataError);
         }
-    };
-    db.read_write(move |dao| {
-        for chunk in interest_data.chunks(WRITE_CHUNK_SIZE) {
-            for (url_hash, interest) in chunk {
-                dao.add_url_interest(*url_hash, *interest)?;
-            }
-            dao.err_if_interrupted()?;
-        }
-        Ok(())
-    })
+    }
+    Ok(())
 }
 
 /// Fetch the interest data
-fn fetch_interest_data() -> std::io::Result<Vec<(UrlHash, Interest)>> {
-    // TODO: this hack should be replaced with something that fetches from remote settings.
-    // It should ideally check for interruption while fetching the data.
-    let bytes = include_bytes!("../test-data");
-    let mut reader = Cursor::new(&bytes);
+fn fetch_interest_data() -> Result<Vec<(Interest, UrlHash)>> {
+    let rs = Client::new(RemoteSettingsConfig {
+        collection_name: REMOTE_SETTINGS_COLLECTION.to_string(),
+        server: Some(RemoteSettingsServer::Prod),
+        server_url: None,
+        bucket_name: None,
+    })?;
+    let remote_settings_response = rs.get_records()?;
     let mut data = vec![];
+    for record in remote_settings_response.records {
+        let record_fields: RelevancyRecord =
+            serde_json::from_value(serde_json::Value::Object(record.fields.clone()))?;
 
-    // Loop over all possible interests
-    for interest in Interest::all() {
-        // read the count
-        let mut buf = [0u8; 4];
-        reader.read_exact(&mut buf)?;
-        let count = u32::from_le_bytes(buf);
-        for _ in 0..count {
-            let mut url_hash: UrlHash = [0u8; 16];
-            reader.read_exact(&mut url_hash)?;
-            data.push((url_hash, interest));
-        }
+        let attachment_data = match &record.attachment {
+            None => return Err(Error::FetchInterestDataError),
+            Some(a) => rs.get_attachment(&a.location)?,
+        };
+
+        let parsed_attachment_data =
+            serde_json::from_slice::<Vec<RelevancyAttachmentData>>(&attachment_data)?;
+        parse_record(record_fields, parsed_attachment_data, &mut data)?;
     }
     Ok(data)
 }
 
+/// Parse record and extract interest info
+fn parse_record(
```

Comment:
Looks like the record & attachment parsing was already done in `fetch_interest_data()`, what it really does here is processing them. What do you think about moving all that together? Doing so would make `fetch_interest_data()` slightly easier to follow as we don't need to pass that mutable `data` around any more.

```rs
/// Fetch the interest data
fn fetch_interest_data() -> Result<Vec<(Interest, UrlHash)>> {
    let rs = Client::new(RemoteSettingsConfig {
        collection_name: REMOTE_SETTINGS_COLLECTION.to_string(),
        server: Some(RemoteSettingsServer::Prod),
        server_url: None,
        bucket_name: None,
    })?;
    let remote_settings_response = rs.get_records()?;

    remote_settings_response
        .records
        .iter()
        .map(|record| parse_record(record, &rs))
        // This would turn a `Vec<Result<T, E>>` into a `Result<Vec<T>, E>`.
        // See more details https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect   
        .collect::<Result<Vec<_>>>()
        .map(|v| v.into_iter().flatten().collect())
}

/// Parse an RS record and its attachment to extract interest info
fn parse_record(
    record: &RemoteSettingsRecord,
    rs_client: &Client,
) -> Result<Vec<(Interest, UrlHash)>> {
    let record_fields: RelevancyRecord =
        serde_json::from_value(serde_json::Value::Object(record.fields.clone()))?;

    let attachment_data = match &record.attachment {
        None => return Err(Error::FetchInterestDataError),
        Some(a) => rs_client.get_attachment(&a.location)?,
    };

    let parsed_attachment_data =
        serde_json::from_slice::<Vec<RelevancyAttachmentData>>(&attachment_data)?;

    let custom_details = record_fields.record_custom_details;
    let category_code = custom_details.category_to_domains.category_code;

    let mut data = Vec::new();

    for attachment_data in parsed_attachment_data {
        let hash_url = STANDARD
            .decode(attachment_data.domain)
            .map_err(|_| Error::Base64DecodeError("Invalid base64 error".to_string()))?;
        let url_hash = hash_url.try_into().map_err(|_| {
            Error::Base64DecodeError("Base64 string has wrong number of bytes".to_string())
        })?;
        let interest = Interest::try_from(category_code as u32)?;
        data.push((interest, url_hash));
    }
    Ok(data)
}
```

Reply:
oooh we may or may not have done it this way to avoid mocking a remote settings client, for testing 🙃

Reply:
In this version `rs_client` is only used to for `get_attachment`.  I think we could refactor so that `fetch_interest_data()` called `get_attachment` and `parse_record` inputted the `Vec<u8>` result.  In fact, I think that was essentially our 2nd pass at this one.

Reply:
Yup, I figure that was the reason. Though that still leaves the bulk of `fetch_interest_data()` untested, which is not ideal as future RS schema related errors would only be captured in production.

Generics is often used for this in Rust and [this](https://github.com/mozilla/application-services/blob/main/components/suggest/src/rs.rs#L67) is how we handle it in `suggest`. Feel free to file a follow-up ticket for this testing enhancement.

Reply:
>  I think we could refactor so that fetch_interest_data() called get_attachment and parse_record inputted the Vec<u8> result.

Makes sense.